### PR TITLE
Sandbox, LedgerAPI Server: Allow specifying a metric prefix.

### PR DESCRIPTION
**Pull request by gerolf-da**
CHANGELOG_BEGIN
[Sandbox][Ledger Integration Kit]: Metrics can now be exported to graphite with a prefix using the format ``--metrics-reporter=graphite://host:port/prefix``.
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala [hunk 1]
```diff
@@ -34,10 +34,12 @@ object MetricsReporter {
         .build(directory.toFile)
   }
 
-  final case class Graphite(address: InetSocketAddress) extends MetricsReporter {
+  final case class Graphite(address: InetSocketAddress, prefix: Option[String] = None)
+      extends MetricsReporter {
     override def register(registry: MetricRegistry): ScheduledReporter =
       metrics.graphite.GraphiteReporter
         .forRegistry(registry)
+        .prefixedWith(prefix.orNull)
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala:42]: Ah, there it is. I was wondering how to do this all in one place.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala [hunk 2]
```diff
@@ -52,8 +54,25 @@ object MetricsReporter {
       Graphite(new InetSocketAddress(defaultHost, port))
   }
 
-  implicit val metricsReporterRead: Read[MetricsReporter] = Read.reads {
-    _.split(":", 2).toSeq match {
+  implicit val metricsReporterRead: Read[MetricsReporter] = Read.reads { value =>
+    if (value == "console") Console
+    else if (!value.contains("://")) {
+      readLegacy(value)
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala:60]: This is not documented, the options are hidden, and I don't know of anyone using it. I don't think we need to support legacy users.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala [hunk 3]
```diff
@@ -52,8 +54,25 @@ object MetricsReporter {
...
+    } else {
+      val uri = Read.uriRead.reads(value)
+      uri.getScheme match {
+        case "csv" => Csv(Paths.get(uri.getPath))
+        case "graphite" =>
+          val port = if (uri.getPort > 0) uri.getPort else Graphite.defaultPort
+          val address = new InetSocketAddress(uri.getHost, port)
+          val metricPrefix = Some(uri.getPath.stripPrefix("/")).filter(_.nonEmpty)
+          Graphite(address, metricPrefix)
+      }
```
Comment (SamirTalwar-DA) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala:70]: Can you add a `case _` that throws a reasonable exception, like in the `readLegacy` function below?

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala (hunk 2 above)
Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala:60]: I was mostly trying to not break any existing deployments.

Comment (stefanobaghino-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/configuration/MetricsReporter.scala]: Well, I think you actually knew at least one user. :wink: https://github.com/DACH-NY/spider-on-x/pull/34

**Pull request merged**